homeassistant/components/light/yeelight.py

```diff
@@ -6,157 +6,311 @@
+DEFAULT_TRANSITION = 350
+
+CONF_SAVE_ON_CHANGE = "save_on_change"
+CONF_MODE_MUSIC = "use_music_mode"
 
-SUPPORT_YEELIGHT = (SUPPORT_BRIGHTNESS | SUPPORT_RGB_COLOR |
-                    SUPPORT_COLOR_TEMP)
+DOMAIN = 'yeelight'
 
-DEVICE_SCHEMA = vol.Schema({vol.Optional(CONF_NAME): cv.string, })
+DEVICE_SCHEMA = vol.Schema({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_TRANSITION, default=DEFAULT_TRANSITION): cv.positive_int,
+    vol.Required(CONF_MODE_MUSIC, default=False): cv.boolean,
+    vol.Required(CONF_SAVE_ON_CHANGE, default=True): cv.boolean,
+})
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
     {vol.Optional(CONF_DEVICES, default={}): {cv.string: DEVICE_SCHEMA}, })
 
+SUPPORT_YEELIGHT_RGB = (SUPPORT_RGB_COLOR |
+                        SUPPORT_COLOR_TEMP)
+
+SUPPORT_YEELIGHT = (SUPPORT_BRIGHTNESS |
+                    SUPPORT_TRANSITION |
+                    SUPPORT_FLASH)
+
+
+def _cmd(func):
+    """A wrapper to catch exceptions from the bulb."""
+    def _wrap(self, *args, **kwargs):
+        import yeelight
+        try:
+            _LOGGER.debug("Calling %s with %s %s", func, args, kwargs)
+            return func(self, *args, **kwargs)
+        except yeelight.BulbException as ex:
+            _LOGGER.error("Error when calling %s: %s", func, ex)
+
+    return _wrap
+
 
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Setup the Yeelight bulbs."""
     lights = []
     if discovery_info is not None:
-        device = {'name': discovery_info['hostname'],
-                  'ipaddr': discovery_info['host']}
-        lights.append(YeelightLight(device))
+        _LOGGER.debug("Adding autodetected %s", discovery_info['hostname'])
+
+        # not using hostname, as it seems to vary.
+        name = "yeelight_%s_%s" % (discovery_info["device_type"],
+                                   discovery_info["properties"]["mac"])
+        device = {'name': name, 'ipaddr': discovery_info['host']}
+
+        lights.append(YeelightLight(device, DEVICE_SCHEMA({})))
     else:
         for ipaddr, device_config in config[CONF_DEVICES].items():
+            _LOGGER.debug("Adding configured %s", device_config[CONF_NAME])
+
             device = {'name': device_config[CONF_NAME], 'ipaddr': ipaddr}
-            lights.append(YeelightLight(device))
+            lights.append(YeelightLight(device, device_config))
 
     add_devices(lights)
 
 
 class YeelightLight(Light):
     """Representation of a Yeelight light."""
 
-    def __init__(self, device):
+    def __init__(self, device, config):
         """Initialize the light."""
-        import pyyeelight
-
+        self.config = config
         self._name = device['name']
         self._ipaddr = device['ipaddr']
-        self.is_valid = True
-        self._bulb = None
-        self._state = None
-        self._bright = None
+
+        self._supported_features = SUPPORT_YEELIGHT
+        self._available = False
+        self.__bulb = None
```

Comment:
Can we rename this to something like `self._bulb_device` so it is more distinct?